crates/sui/src/sui_commands.rs

```diff
@@ -430,6 +509,221 @@ impl SuiCommand {
+    }
+
+    let mut fullnode_url = sui_config::node::default_json_rpc_address();
+    fullnode_url.set_port(fullnode_rpc_port);
+
+    if no_full_node {
+        swarm_builder = swarm_builder.with_fullnode_count(0);
+    } else {
+        swarm_builder = swarm_builder
+            .with_fullnode_count(1)
+            .with_fullnode_rpc_addr(fullnode_url);
+    }
+
+    let mut swarm = swarm_builder.build();
+    swarm.launch().await?;
+    // Let nodes connect to one another
+    tokio::time::sleep(tokio::time::Duration::from_secs(2)).await;
+    info!("Cluster started");
+
+    // the indexer requires a full url
+    let fullnode_url = format!("http://{}", fullnode_url);
+    info!("Fullnode URL: {}", fullnode_url);
+    let pg_address = format!("postgres://{pg_user}:{pg_password}@{pg_host}:{pg_port}/{pg_db_name}");
+
+    if with_indexer {
+        let indexer_address = format!("0.0.0.0:{}", indexer_rpc_port);
+        // Start in writer mode
+        start_test_indexer::<diesel::PgConnection>(
+            Some(pg_address.clone()),
+            fullnode_url.clone(),
+            ReaderWriterConfig::writer_mode(None),
+            data_ingestion_path.clone(),
+        )
+        .await;
+        info!("Indexer in writer mode started");
+
+        // Start in reader mode
+        start_test_indexer::<diesel::PgConnection>(
+            Some(pg_address.clone()),
+            fullnode_url.clone(),
+            ReaderWriterConfig::reader_mode(indexer_address.to_string()),
+            data_ingestion_path,
+        )
+        .await;
+        info!("Indexer in reader mode started");
+    }
+
+    if with_graphql {
+        let graphql_connection_config = ConnectionConfig::new(
+            Some(graphql_port),
+            Some(graphql_host),
+            Some(pg_address),
+            None,
+            None,
+            None,
+        );
+        start_graphql_server_with_fn_rpc(
+            graphql_connection_config,
+            Some(fullnode_url.clone()),
+            None,
+        )
+        .await;
+        info!("GraphQL started");
+    }
+    if with_faucet {
+        let faucet_key = swarm.config_mut().account_keys.swap_remove(0);
+        let config_dir = tempdir()?.into_path();
+        let prom_registry = prometheus::Registry::new();
+        let config = FaucetConfig::default();
+        let simple_faucet = SimpleFaucet::new(
+            new_wallet_context_for_faucet(faucet_key, config_dir.clone(), fullnode_url.clone())?,
+            &prom_registry,
+            config_dir.join("faucet.wal").as_path(),
+            config,
+        )
+        .await
+        .unwrap();
+
+        let local_faucet_client = Arc::new(LocalFaucetClient::new(simple_faucet));
+        let app_state = Arc::new(AppState::new(local_faucet_client));
+
+        start_faucet(app_state, faucet_port).await?;
+    }
+
+    let mut interval = tokio::time::interval(std::time::Duration::from_secs(3));
+    let mut unhealthy_cnt = 0;
+    loop {
```

Comment:
How do we deal with teardown if the user interrupts the running localnet -- will it cleanly bring down all the services? I couldn't figure out from looking at the code how this would be handled, but maybe it just works because each service deals with it on its own.

Reply:
I have not changed anything from what it used to be, so I think the answer is that nothing much gets cleaned up properly (as before). The nodes have a `.stop()` function we could use and which I think it does not get used. The `start_test_indexer` functions seem to instantiate a new CancellationToken when this function is called. 
I guess faucet is the one to be revised to support this.

Reply:
If this isn't a regression from how `sui-test-validator` operates then it's not a blocker for this PR, but we should come back and clean it up -- otherwise the teardown could corrupt various service's states and ruin people's day.

Reply:
Let's revisit this in a follow up PR.